### PR TITLE
Newton system assembly

### PR DESCRIPTION
**Pull request by MFraters**
Is the structure of this newton assembly correct? I didn't change the terms yet, so we can first see if this is the structure we want to use.

### CODE REVIEW

File: source/simulator/assembly.cc [hunk 1]
```diff
@@ -1227,6 +1227,206 @@ namespace aspect
         }
     };
 
+    /**
+     * A class for the definition of functions that implement the
+     * non-linear system terms for the newton compressible or
+     * incompressible equations.
+     */
+    template <int dim>
+    class NewtonSystem : public aspect::internal::Assembly::Assemblers::AssemblerBase<dim>,
```
Comment (bangerth) [inline source/simulator/assembly.cc]: call it `NewtonStokes` or similar to make clear that you're only applying the Newton solver to this part of the problem.

File: source/simulator/assembly.cc [hunk 2]
```diff
@@ -1227,6 +1227,206 @@ namespace aspect
...
+      public SimulatorAccess<dim>
+    {
+      public:
+        void
+        local_assemble_newton_preconditioner (const double                                             pressure_scaling,
```
Comment (bangerth) [inline source/simulator/assembly.cc]: I'd still call it `local_assemble_stokes_preconditioner`. The "Newton" part is already taken care of by the class name.

File: source/simulator/assembly.cc [hunk 3]
```diff
@@ -1227,6 +1227,206 @@ namespace aspect
...
+                                                   .phi_p[j]))
+                                               * scratch.finite_element_values.JxW(
+                                                 q);
+            }
+        }
+
+
+
+        void
+        local_assemble_newton_system_compressible (const double                                     pressure_scaling,
```
Comment (bangerth) [inline source/simulator/assembly.cc]: same here and below.

File: source/simulator/assembly.cc [hunk 4]
```diff
@@ -1227,6 +1227,206 @@ namespace aspect
...
+                  scratch.phi_p[k] = scratch.finite_element_values[introspection.extractors.pressure].value (k, q);
+                  if (rebuild_stokes_matrix)
+                    {
+                      scratch.grads_phi_u[k] = scratch.finite_element_values[introspection.extractors.velocities].symmetric_gradient(k,q);
+                      scratch.div_phi_u[k]   = scratch.finite_element_values[introspection.extractors.velocities].divergence (k, q);
+                    }
+                }
+
+              // Viscosity scalar
+              const double eta = (rebuild_stokes_matrix
```
Comment (bangerth) [inline source/simulator/assembly.cc]: If you assembler the Newton system, you will *always* have to rebuild the Stokes matrix. So here and below, simplify the code by assuming that the variable is true.

**Comment by bangerth** [comment]
Yes, the general structure looks right.

**Comment by bangerth** [comment]
In reference to #733.

**Comment by MFraters** [comment]
I think that I now got the assembly working for the incompressible case. I also added the derivative structure to this commit and when I run it, it now actually only crashes at the second nonlinear iteration of the crustal model 2d! :)

**Comment by bangerth** [comment]
Alright, that's already progress :-)

**Comment by MFraters** [comment]
Added the structure for a Newton Stokes system. Still have to look at the solver.cc

**Comment by MFraters** [comment]
@bangerth Could you take a look at the current changes? I am facing the problem that the divergence of the velocity I calculate for computing the right hand side in the assembly becomes nan (for i >=10). Am I computing it wrong? I used the test with the nonlinear channel flow which is included in this pull request. It is basically a nonlinear version of the 2d poiseuille flow with pressure boundary conditions.

**Comment by MFraters** [comment]
Here is a working version of the newton method with aspect. It isn't fully clean yet, it I thought it would be better to first discuss the large scale structure before cleaning the small scale. As you can see, the newton iteration converges quadratically near the solution for the nonlinear channel flow test. 

There are two comments about the current code/tests:
1. For this test both the newton iteration and the picard iteration can't converge past 1e-8. Not sure where this comes from. Any ideas?
2. The newton method seems to be very sensitive to the linear tolerance. I tried to solve this with with the Eisenstat Walker method (1994 & 1996, a method to prevent oversolving. Also looked at a report from Delft: http://ta.twi.tudelft.nl/nw/users/vuik/papers/Ide11LV.pdf), and potentially save a lot of run time. But I didn't get it to work yet (got problems with AztecOO::Iterate error code -2: numerical breakdown). I commented out those parts, but there are still visible in the code.

**Comment by MFraters** [comment]
@bangerth I need some help with the following: I have found a special case where the finite difference and analytical solution do not match. And although I think I understand why that is the case, I am not sure which of the two is the correct solution.

In short the problem is that for all common cases I found that the finite difference and the analytical solution are exactly the same. But with a strain rate in which the diagonal entries are the same the analytical solution gives -0 and the finite difference gives results on the order of the other strain rates (~-1e+20).

Here is the relevant code (There is also a minimum to edot_ii in the real code.):
   const double edot_ii = 2 * std::sqrt(0.5*deviator(in.strain_rate[i])*deviator(in.strain_rate[i]));
   const double stress_exponent_inv = (1./stress_exponent[c]);
   composition_viscosities[c] = std::pow(prefactor[c],-stress_exponent_inv) * std::pow(edot_ii,stress_exponent_inv-1);
   composition_viscosities_derivatives[c] = 2 * (stress_exponent_inv-1) * composition_viscosities[c] * (1/(edot_ii*edot_ii)) * deviator(in.strain_rate[i]);

I think this happens for the following reason: In the analytical equation, the deviator make the diagonal then zero, and the analytical solution then also ends up with zero entries on the diagonal. The finite difference method first gives the original strain rate tensor to the material model, and the deviator makes the diagonal zero. But then the finite difference method perturbs only one of the diagonal entries, which make the diagonal entries not equal any more and as a consequence the diagonal of the deviator will not contain zero entries any more, and I think this lead to a viscosity change on the order of what normally happens.

I hope my explanation is clear enough, if not, please let me know and I will try to write it out a bit more. I hope you know which solution to trust.

Thanks!

**Comment by MFraters** [comment]
This pull request has been superseded.

**Pull request closed without merging**